// parse-stack/parse-stack.js
;(function(){

/**
 * Require the given path.
 *
 * @param {String} path
 * @return {Object} exports
 * @api public
 */

function require(path, parent, orig) {
  var resolved = require.resolve(path);

  // lookup failed
  if (null == resolved) {
    orig = orig || path;
    parent = parent || 'root';
    var err = new Error('Failed to require "' + orig + '" from "' + parent + '"');
    err.path = orig;
    err.parent = parent;
    err.require = true;
    throw err;
  }

  var module = require.modules[resolved];

  // perform real require()
  // by invoking the module's
  // registered function
  if (!module._resolving && !module.exports) {
    var mod = {};
    mod.exports = {};
    mod.client = mod.component = true;
    module._resolving = true;
    module.call(this, mod.exports, require.relative(resolved), mod);
    delete module._resolving;
    module.exports = mod.exports;
  }

  return module.exports;
}

/**
 * Registered modules.
 */

require.modules = {};

/**
 * Registered aliases.
 */

require.aliases = {};

/**
 * Resolve `path`.
 *
 * Lookup:
 *
 *   - PATH/index.js
 *   - PATH.js
 *   - PATH
 *
 * @param {String} path
 * @return {String} path or null
 * @api private
 */

require.resolve = function(path) {
  if (path.charAt(0) === '/') path = path.slice(1);

  var paths = [
    path,
    path + '.js',
    path + '.json',
    path + '/index.js',
    path + '/index.json'
  ];

  for (var i = 0; i < paths.length; i++) {
    var path = paths[i];
    if (require.modules.hasOwnProperty(path)) return path;
    if (require.aliases.hasOwnProperty(path)) return require.aliases[path];
  }
};

/**
 * Normalize `path` relative to the current path.
 *
 * @param {String} curr
 * @param {String} path
 * @return {String}
 * @api private
 */

require.normalize = function(curr, path) {
  var segs = [];

  if ('.' != path.charAt(0)) return path;

  curr = curr.split('/');
  path = path.split('/');

  for (var i = 0; i < path.length; ++i) {
    if ('..' == path[i]) {
      curr.pop();
    } else if ('.' != path[i] && '' != path[i]) {
      segs.push(path[i]);
    }
  }

  return curr.concat(segs).join('/');
};

/**
 * Register module at `path` with callback `definition`.
 *
 * @param {String} path
 * @param {Function} definition
 * @api private
 */

require.register = function(path, definition) {
  require.modules[path] = definition;
};

/**
 * Alias a module definition.
 *
 * @param {String} from
 * @param {String} to
 * @api private
 */

require.alias = function(from, to) {
  if (!require.modules.hasOwnProperty(from)) {
    throw new Error('Failed to alias "' + from + '", it does not exist');
  }
  require.aliases[to] = from;
};

/**
 * Return a require function relative to the `parent` path.
 *
 * @param {String} parent
 * @return {Function}
 * @api private
 */

require.relative = function(parent) {
  var p = require.normalize(parent, '..');

  /**
   * lastIndexOf helper.
   */

  function lastIndexOf(arr, obj) {
    var i = arr.length;
    while (i--) {
      if (arr[i] === obj) return i;
    }
    return -1;
  }

  /**
   * The relative require() itself.
   */

  function localRequire(path) {
    var resolved = localRequire.resolve(path);
    return require(resolved, parent, path);
  }

  /**
   * Resolve relative to the parent.
   */

  localRequire.resolve = function(path) {
    var c = path.charAt(0);
    if ('/' == c) return path.slice(1);
    if ('.' == c) return require.normalize(p, path);

    // resolve deps by returning
    // the dep in the nearest "deps"
    // directory
    var segs = parent.split('/');
    var i = lastIndexOf(segs, 'deps') + 1;
    if (!i) i = 0;
    path = segs.slice(0, i + 1).join('/') + '/deps/' + path;
    return path;
  };

  /**
   * Check if module is defined at `path`.
   */

  localRequire.exists = function(path) {
    return require.modules.hasOwnProperty(localRequire.resolve(path));
  };

  return localRequire;
};
require.register("parse-stack/lib/parse-stack.js", function(exports, require, module){
// Generated by CoffeeScript 1.6.3
/*
Copyright 2013 Simon Lydell

This file is part of parse-stack.

parse-stack is free software: you can redistribute it and/or modify it under the terms of the GNU
Lesser General Public License as published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

parse-stack is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
General Public License for more details.

You should have received a copy of the GNU Lesser General Public License along with parse-stack. If
not, see <http://www.gnu.org/licenses/>.
*/

var ensureType, formats, newline, parseStack;

formats = [/^\x20+at\x20(?:([^(]+)\x20\()?(\S*):(\d+):(\d+)\)?$/, /^([^@]*)@(\S*):(\d+)$/];

newline = /\r\n|\r|\n/;

parseStack = function(error) {
  var columnNumber, filepath, format, index, lineNumber, match, name, prelude, stack, stackLine, stackLines, _i, _j, _len, _len1, _ref, _ref1, _results;
  if (typeof error.stack !== "string") {
    return null;
  }
  stack = error.stack;
  prelude = error.toString();
  if (stack.slice(0, prelude.length) === prelude) {
    stack = stack.slice(prelude.length);
  }
  stackLines = stack.split(newline).filter(function(stackLine) {
    return stackLine !== "";
  });
  for (_i = 0, _len = formats.length; _i < _len; _i++) {
    format = formats[_i];
    match = stackLines[0].match(format);
    if (match) {
      break;
    }
  }
  if (!match) {
    throw new Error("Unkown stack trace format:\n" + stack);
  }
  _results = [];
  for (index = _j = 0, _len1 = stackLines.length; _j < _len1; index = ++_j) {
    stackLine = stackLines[index];
    _ref1 = (_ref = stackLine.match(format)) != null ? _ref : [], match = _ref1[0], name = _ref1[1], filepath = _ref1[2], lineNumber = _ref1[3], columnNumber = _ref1[4];
    if (!match) {
      throw new Error("Unknown stack trace formatting on stack line " + (index + 1) + ":\n" + stackLine);
    }
    name = ensureType(String, name);
    filepath = ensureType(String, filepath);
    lineNumber = ensureType(Number, lineNumber);
    columnNumber = ensureType(Number, columnNumber);
    _results.push({
      name: name,
      filepath: filepath,
      lineNumber: lineNumber,
      columnNumber: columnNumber
    });
  }
  return _results;
};

ensureType = function(type, value) {
  if (value) {
    return type(value);
  } else {
    return void 0;
  }
};

switch (false) {
  case !(typeof define === "function" && define.amd):
    define(parseStack);
    break;
  case typeof module !== "object":
    module.exports = parseStack;
    break;
  default:
    this.parseStack = parseStack;
}

});
require.alias("parse-stack/lib/parse-stack.js", "parse-stack/index.js");if (typeof exports == "object") {
  module.exports = require("parse-stack");
} else if (typeof define == "function" && define.amd) {
  define(function(){ return require("parse-stack"); });
} else {
  this["parse-stack"] = require("parse-stack");
}})();
